### DISCUSSION TITLE
Make $queryTags of ElementQuery::getCacheTags() and Element::getCacheTags() overridable via event

### DISCUSSION THREAD

**Discussion by d--j** [craftcms/cms#11171] [Ideas] [answered]
I have some behaviors that augment Entry and EntryQuery. It would be great when I also could add some more cache tags to the Entry and the EntryQuery so that the query cacheing and invalidation can be optimized. 

E.g. I have a project with several thousand entries in a section where there are quite frequent modifications to those entries (automated sync of an external system). Since queries are made primarily via the external ID they almost all get the `*` catch-all tag and thus get invalidated on every entry save. It would be good when I could add the external ID as a new cache tag.

Replace

https://github.com/craftcms/cms/blob/5fbbe7523513c6a4fd2bf7a4be64799219dbfb75/src/elements/db/ElementQuery.php#L1882-L1885

with

```php
            $event = new DefineCacheTagsEvent(['tags' => $queryTags]);
            $this->trigger(self::EVENT_DEFINE_CACHE_TAGS, $event);
            $queryTags = $event->tags;

            foreach ($queryTags as $tag) {
                $this->_cacheTags[] = "element::$this->elementType::$tag";
            }
```

and replace

https://github.com/craftcms/cms/blob/5fbbe7523513c6a4fd2bf7a4be64799219dbfb75/src/base/Element.php#L2571-L2574

with

```php
    public function getCacheTags(): array
    {
        $event = new DefineCacheTagsEvent(['tags' => $this->cacheTags()]);
        $this->trigger(self::EVENT_DEFINE_CACHE_TAGS, $event);
        return $event->tags;
    }

    protected function cacheTags(): array
    {
        return [];
    }

```

rename all `getCacheTags`-methods of Element descendants to `cacheTags`

**Comment by brandonkelly** [comment, marked as answer]
Just added `EVENT_DEFINE_CACHE_TAGS` events to `craft\base\Element::getCacheTags()` and `craft\elements\db\ElementQuery::getCacheTags()` for Craft 4.1 🎉  (18f1bb2ddb01af22bbd4d8acc31b978af67ee0d8)

Element types will need to start extending (the new) `protected function cacheTags()` instead of `public function getCacheTags()` for the event to trigger.

**Reply by brandonkelly** [reply]
Craft 4.1 is out with that new event!

**Reply by d--j** [reply]
Cool! Thanks for the quick implementation of this feature!

For anyone interested, here is how you wire this up:

```php
use craft\base\Element;
use craft\elements\db\EntryQuery;
use craft\elements\db\ElementQuery;
use craft\elements\Entry;
use craft\events\DefineValueEvent;
use yii\base\Event;

Event::on(Entry::class, Element::EVENT_DEFINE_CACHE_TAGS, static function (DefineValueEvent $e) {
    /** @var Entry $entry */
    $entry = $e->sender;
    $id    = $entry->getFieldValue('fieldPondusId');
    if (! empty($id)) {
        $e->value[] = "pondus:$id";
    }
});
Event::on(EntryQuery::class, ElementQuery::EVENT_DEFINE_CACHE_TAGS, static function (DefineValueEvent $e) {
    /** @var EntryQuery $query */
    $query = $e->sender;
    /** @var CustomFieldBehavior $behaviour */
    $behaviour = $query->getBehavior('customFields');
    /** @var string|int|string[]|int[] $pondusId */
    $pondusId = $behaviour->fieldPondusId;
    if (! empty($pondusId)) {
        $e->value = array_merge($e->value, array_map(fn ($id) => "pondus:$id", (array)$pondusId));
    }
});
```


This will invalidate `"pondus:$externalId"` cache tags whenever the entry gets saved and will attach these cache tags to queries, that lookup via the external id e.g. `Entry::find()->section('aSection')->fieldPondusId(15)` will get the cache tag `pondus:15` attached.


You might need to do special handling if the external ID of your entries can change. (E.g. an entry was associated with external id 16 but the user changed it to external id 15). This cannot happen in my case (`fieldPondusId` is a read only field) so I am lucky.

**Comment by khalwat** [comment]
Related PR: https://github.com/craftcms/cms/pull/11617
